### ISSUE TITLE
Peer-to-Peer extension store, or at the very least, decentralized extension store.

### ISSUE THREAD

**Issue by KSXGitHub** [lapce/lapce#1850]
I haven't find any documentation that mentions how Lapce Plugins are distributed. I hope it's going to be peer-to-peer.

## Is your feature request related to a problem? Please describe.

An editor that works should keep working. Atom was shut down recently, and whilst the core editor still work, it's extension ecosystem doesn't.

A FOSS editor shouldn't lock user in a centralize ecosystem. VS Code locks user in a proprietary Microsoft store, and whilst Open VSX can work, it cannot be compared to the Microsoft store.

Protect the ecosystem against DDoS attacks, NSP blocks, and the likes.

## Describe the solution you'd like

**Level 1:** Protocol agnostic content-addressable system.
* Packages are primarily referred to by the hash of its content.
* All packages are GPG signed.
* There may be an online registry that map human-readable aliases to hash addresses.
* There may be a local file that is the subset of aforementioned online registry.

**Level 2:** Develop or adopt a P2P protocol.
* It should work like IPFS or torrent.
* It should not depend on a single server.
* If a machine that shares my LAN has package X, I should be able to download that package X.

**Comment by panekj** [comment]
> I haven't find any documentation that mentions how Lapce Plugins are distributed.

We have plugin registry at https://plugins.lapce.dev backed by https://github.com/lapce/lapce-volts

> **Level 1:** Protocol agnostic content-addressable system.

It should not be protocol agnostic, it should be a well defined spec that can be reusable by others if they wish to implement registry themselves.

> All packages are GPG signed.

That's just not possible at all. It would have to either be signed on registry which defeats whole purpose of signing or you drive people away because of how GPG is PITA and unusable.
I'm not opposed to allowing plugins to be signed and show appropriate notice(+details) when they are, but requiring for every package to be signed is just not viable at all.

> * There may be an online registry that map human-readable aliases to hash addresses.
> * Packages are primarily referred to by the hash of its content.

I don't see a point in that, namespace-based identification based on centralised registry is perfectly fine.
It would also unnecessarily confuse users and cause issues when/if plugin is still same version but shows as different.

> **Level 2:** Develop or adopt a P2P protocol.
> 
> * It should work like IPFS or torrent.

No, it should not, peer-based protocols are unnecessary, unreliable and unviable for such tasks.
- Direct connection between user instances should never be a thing unless on local network.
- It would not provide guarantee that plugin is available.
- Server-side peer connections introduce moderation issues that we are not interested in dealing with.

> * It should not depend on a single server.

It does not, we don't use a single server, although the registry is centralised in specific location currently (but it's fine for current use).
If you mean it should not depend on a single *registry*, it's up to you if you do not want to use it.

> * If a machine that shares my LAN has package X, I should be able to download that package X.

This introduces code complexity, maintenance burden and security implications for no gain at all as the situation where you would install plugin on the local network would be extremely rare.

**Comment by yashpalgoyal1304** [comment]
_self hiding as offtopic_

----
Earth, Solar System, Orion Arm, Milky Way


oh lolol...

**Comment by khimaros** [comment]
@panekj great answers here and happy to see that the server is open source. is there any way to download the full database dump from the current repository?

**Comment by panekj** [comment]
> @panekj great answers here and happy to see that the server is open source. is there any way to download the full database dump from the current repository?

I'm unsure what you mean by "full database dump". We are not going to distribute the content of plugins.lapce.dev database since it contains sensitive user data such as private tokens to interact with registry API.

**Comment by khimaros** [comment]
i suppose i meant a dump of the public extension data. whatever is needed to make a full mirror of the plugin registry. this also assumes the registry endpoint can be configured in lapce.

**Comment by panekj** [comment]
> i suppose i meant a dump of the public extension data.

It's not currently possible to do that, neither we're going to allow that. Servers and traffic cost money.
It's also not possible to change registry used.

**Comment by khimaros** [comment]
wouldn't it save money in the long run to support mirrors?

**Comment by panekj** [comment]
No, because mirrors would be used only by users who configured Lapce to use those.

**Comment by khimaros** [comment]
you seem to be describing a bootstrapping problem. if the capability to create and configure mirrors is not present, then it will never be possible to reduce server costs via mirroring. consider the ideal long term outcome and the steps to get there will become clearer.

**Comment by panekj** [comment]
> you seem to be describing a bootstrapping problem.

no...

> if the capability to create and configure mirrors is not present, then it will never be possible to reduce server costs via mirroring.

It will never be possible because majority of users will not configure different mirror, neither will care about this.

> consider the ideal long term outcome and the steps to get there will become clearer.

I actually did that, and we're already on that path.

**Comment by Yiniau** [comment]
> It will never be possible because majority of users will not configure different mirror, neither will care about this.

mirrors are necessary for me; in other words, necessary for most people live in China.

**Comment by Octorine** [comment]
If Lapce takes off, I'd like to be able to use it at work someday, but my company (and many others) is never going to allow me to use a plugin registry.  For them to allow installing lapce, they would have to have their own registry where every plugin has been  vetted by their security team.

I think Rust had a similar issue with cargo where some companies couldn't use it because of crates.io, which is why cargo vendor exists now.

We're probably a long time away from this mattering, but if Lapce is successful, it'll eventually need to support something like this.
